debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/PostgresReplicationConnection.java

```diff
@@ -198,12 +205,82 @@ public ReplicationStream startStreaming(Long offset) throws SQLException, Interr
+
+    @Override
+    public void createReplicationSlot() throws SQLException {
+        // note that some of these options are only supported in pg10 or newer, additionally
+        // the options are not yet exported by the jdbc api wrapper, therefore, we just do this ourselves
+        // but eventually this should be moved back to the jdbc API
+        // see https://github.com/pgjdbc/pgjdbc/issues/1305
+
+        LOGGER.debug("Creating new replication slot '{}' for plugin '{}'", slotName, plugin);
+        String tempPart = "";
+        String exportPart = "";
+        Boolean isPg10 = isPg10OrGreater();
+        if ((dropSlotOnClose || exportSnapshot) && !isPg10OrGreater()) {
+            LOGGER.warn("A slot marked as temporary or with an exported snapshot was created, but not on a supported version of Postgres, ignoring!");
+        }
+        if (useTemporarySlot()) {
+            tempPart = "TEMPORARY";
+        }
+        if (exportSnapshotOnCreate()) {
+            exportPart = "EXPORT_SNAPSHOT";
+        }
+        try (Statement stmt = pgConnection().createStatement()) {
+            String createCommand = String.format(
+                    "CREATE_REPLICATION_SLOT %s %s LOGICAL %s %s",
+                    slotName,
+                    tempPart,
+                    plugin.getPostgresPluginName(),
+                    exportPart
+            );
+            LOGGER.info("Creating replication slot with command {}", createCommand);
+            stmt.execute(createCommand);
+            // when we are in pg10, we can parse the slot creation info, otherwise, it returns
+            // nothing
+            if (isPg10) {
+                this.slotCreationInfo = parseSlotCreation(stmt.getResultSet());
+            }
+        }
+    }
+
+    @Override
+    public Optional<SlotCreationResult> getSlotCreationResult() {
+        return Optional.ofNullable(slotCreationInfo);
+    }
+
     protected PgConnection pgConnection() throws SQLException {
         return (PgConnection) connection(false);
     }
 
+    private SlotCreationResult parseSlotCreation(ResultSet rs) {
+        try {
+            if (rs.next()) {
+                String slotName = rs.getString("slot_name");
+                String startPoint = rs.getString("consistent_point");
+                String snapName = rs.getString("snapshot_name");
+                String pluginName = rs.getString("output_plugin");
+
+                return new SlotCreationResult(slotName, startPoint, snapName, pluginName);
+            }
+            else {
+                throw new ConnectException("expected response to create_replication_slot");
+            }
+        }
+        catch (SQLException ex) {
+            throw new ConnectException("unable to parse create_replication_slot", ex);
+        }
+    }
+
     private ReplicationStream createReplicationStream(final LogSequenceNumber lsn) throws SQLException, InterruptedException {
         PGReplicationStream s;
+        initConnection();
```

Comment:
I think this might be a problem; see the two sub-sequent calls to `initReplicationSlot()` later in this method. They are in try/catch blocks do work-around issues with different versions of wal2json which might reject the slot options we pass. We wouldn't get to these catch blocks any longer due to this call here. This is all a bit messy unfortunately.

That said, what's the need for `initConnection()` to begin with? That's not quite clear to me yet?

Reply:
This actually is currently wrong (about to push up a commit to fix it) but not quite for the reasons listed.

The exceptions below happen when **starting** the slot, not when creating the slot. When you create a slot, you simply specify the plugin, but you don't pass any options. The `initConnection` makes sure the slot exists and isn't actively being used, and also fetches the defaultStartingPosition (which is what is wrong here, we need to call initConnection earlier).

So all his code should be safe to do here